modules/battery_manager/php/testcontroller.class.inc

```diff
@@ -0,0 +1,251 @@
+<?php
+/**
+ * This file contains the Test Controller.
+ *
+ * PHP Version 7
+ *
+ * @category Module
+ * @package  Battery_Manager
+ * @author   Henri Rabalais <henri.rabalais@mcin.ca>
+ * @license  http://www.gnu.org/licenses/gpl-3.0.txt GPLv3
+ * @link     https://www.github.com/aces/Loris/
+ */
+namespace LORIS\battery_manager;
+use \Psr\Http\Message\ServerRequestInterface;
+use \Psr\Http\Server\RequestHandlerInterface;
+use \Psr\Http\Message\ResponseInterface;
+
+/**
+ * Main class for managing Test Instances
+ *
+ * Handles requests for retrieving and saving Test Instances.
+ * Allows users to add, activate, and deactivate entries in the test battery.
+ *
+ * PHP Version 7
+ *
+ * @category Module
+ * @package  Battery_Manager
+ * @author   Henri Rabalais <henri.rabalais@mcin.ca>
+ * @license  http://www.gnu.org/licenses/gpl-3.0.txt GPLv3
+ * @link     https://www.github.com/aces/Loris/
+ */
+class TestController implements RequestHandlerInterface
+{
+
+    /**
+     * Returns true if user has access to this endpoint.
+     *
+     * @param \User $user The user whose access is being checked
+     *
+     * @return bool
+     */
+    function _hasAccess(\User $user) : bool
+    {
+        return true;
+    }
+
+    /**
+     * This function passes the request to the handler. This is necessary since
+     * the Endpoint bypass the Module class.
+     *
+     * XXX: This function should be extracted to a parent class.
+     *
+     * @param ServerRequestInterface  $request The PSR7 request.
+     * @param RequestHandlerInterface $handler The request handler.
+     *
+     * @return ResponseInterface The outgoing PSR7 response.
+     */
+    public function process(
+        ServerRequestInterface $request,
+        RequestHandlerInterface $handler
+    ) : ResponseInterface {
+        return $handler->handle($request);
+    }
+
+    /**
+     * This acts as an Ajax enpoint that handles all action requests from the
+     * Battery Manager Module.
+     *
+     * @param ServerRequestInterface $request The incoming PSR7 request
+     *
+     * @return ResponseInterface The outgoing PSR7 response
+     */
+    public function handle(ServerRequestInterface $request) : ResponseInterface
+    {
+        $db     = \Database::singleton();
+        $user   = $request->getAttribute('user');
+        $method = $request->getMethod();
+        $test   = json_decode($request->getBody()->getContents(), true);
```

Comment:
Why not `new TestInstance(json_decode($request->getBody()->getContents(), true));
you would be able to type hint your private functions with TestInstance instead of array.
Moving `toSQL` from TestController to TestInstance seems suitable as well.

Reply:
Okay